Added section on creating from templates

bfallonf opened the pull request:
Very much a WIP. Not sure if all the correct information is here.

@bparees @smarterclayton I asked @adellape a few questions about this today and he gave me a little more context. I'm not sure if "Project" is the thing that's actually being created from templates here. I can't find anything particularly indicative in my notes.

I'm particularly interested why this is explaining the same process as outlined later in the same file under the "Generating a Configuration" section... At least, the process Ben explained to me via email is essentially the same here.

bparees commented on dev_guide/templates.adoc:
```diff
@@ -51,8 +51,16 @@ A template describes a set of resources intended to be used together that can be
+
+You can create a project from a template that has been uploaded to your OpenShift instance. After a template has been uploaded, you can create a new project from either the Management Console or the CLI.
+
+On the Management Console, click on the Create button on the project page. It will then ask you to choose the template to create from, or upload a new template.
```
i don't think we offer to upload a template today, do we?

jcantrill replied:
@bparees we do not offer an 'upload' function unless this refers to a 'osc create -f <template>'.  I have considered adding some functionality to eclipse which would allow you to 'create app' from a template that you select from disk

bparees commented on dev_guide/templates.adoc:
```diff
@@ -51,8 +51,16 @@ A template describes a set of resources intended to be used together that can be
 
 ====
 
+== Creating a Project From a Template
```
this whole section has a confusing flow.  for starters you don't create projects from templates, projects are containers for "stuff" and templates create some stuff.  Some stuff in a project might be created from a template, some might be created in other ways.

secondly, there are a few conflated flows here:
1) loading a template into the openshift template library so it can be instantiated ("osc create -f mytemplate.json" is the only way to do this today as far as i know)
2) instantiating a template (choose the template via the webconsole, or run "osc process mytemplate | osc create -f -"
3) you can also instantiate a template that's not in the library (ie directly from the filesystem) by running "osc process -f mytemplate.json | osc create -f -"
4) you will also soon be able to instantiate templates via Eclipse, so we wil want at least a call out to docs for that.. @jcantrill can help there.

adellape commented:
@jcantrill Agreed, that section currently exists here: http://docs.openshift.org/latest/dev_guide/console.html#creating-from-template. Bilhar has suggested that that `dev_guide/console.adoc` topic will likely be broken up and absorbed into other relevant topics, so maybe this is an opportunity to just move it here now.

@bfallonf #1 in @bparees' list above would be a new section, like "Loading Templates to a Library" or something. E.g., when you try to browse the template library in the Mgmt Console before loading any, it says: "There are no templates to select from. To add a template to your project run osc create -f <template_file> -n testproject" ("testproject" is the name of the project I was viewing).

The Mgmt Console part of #2 in the list is what you were describing with your WIP content. That flow assumes a template has already been loaded. This is what you could probably just pull/move Cesar's content mentioned above for.

The CLI part of #2 and all of #3 in the list is what is currently in the "Generating a Configuration" section in this topic, though it needs to get updated (it still talks about `osc apply` instead of `create`).

bfallonf commented:
@bparees Ok. I think it looks a lot better now. I've done as @jcantrill and @adellape suggested and moved across the console stuff (cos yes, I remember Bilhar talking about that too), and moved the Generating a Configuration stuff to the relevant part of the doc. I've also added in a small section about uploading the template to the library and reworded a few of the sentences. I have a few more questions though:

1. What exactly is being generated? At first I thought it was a template for a project, but above you've been using the word "stuff" and the docs now use "configuration". I take it a number of things can be created using a template? Is there a way I can find out exactly what can be created here so that I can make a list or something?

2. Is there a difference between `osc create -f` and `osc process -f`. The current CLI docs seem to state the same thing: https://github.com/openshift/openshift-docs/blob/master/cli_reference/basic_cli_operations.adoc#common-operations

and 

https://github.com/openshift/openshift-docs/blob/master/cli_reference/basic_cli_operations.adoc#project-operations

3. @jcantrill Your work on Eclipse is mentioned here. Is there a trello card or something that I can attach myself to so I can see any changes this will need in the future?

Guys, thanks for your help on this!

jcantrill commented:
@bfallonf  In response to your questions:
1) A configuration is being generated from the template.  Early releases it was actually a config kind but that has since been made more generic to just a list.  A template is a parameterized list of openshift resources that once 'processed' will provide you with a list of resources that can be 'applied'/created in openshift.  The resources can be any(at least almost any) valid OpenShift kind (e.g. service, pod, route, buildconfig, etc).
2. 'create' is simply create the resource.  'process' is template specific (I think) and means take the parameters defined in the template and substitute them throughout the list resources in the template and give me back a list of resources that can be created on the server.

In regards to the eclipse work, there are several cards listed here: https://trello.com/b/wqe9UaaZ/openshift-origin-developer-experience.  Specifically, the ones either completed or in play are: https://trello.com/c/vv8XBZky and https://trello.com/c/f2fMXYjc

adellape commented on dev_guide/templates.adoc:
```diff
@@ -10,7 +10,7 @@
 
 == Overview
-A template describes a set of resources intended to be used together that can be customized and processed to produce a configuration. Each template defines a list of parameters that can be modified for consumption by containers. It also defines a set of labels to apply to every resource created by the template.
+A template describes a set of resources intended to be used together that can be customized and processed to produce a configuration. Each template defines a set of parameters that can be modified for consumption by containers, and defines a set of labels to apply to every resource created by the template.
```
It would be nice to be linked here (probably on the word "resources" in the first sentence) to something that helps inform or remind what those resources could be. Maybe the best target right now would be http://docs.openshift.org/latest/architecture/core_objects/overview.html. It's a stub at the moment, but it would at least get them in the right subdir (to then check out `kubernetes_model.adoc` and `openshift_model.adoc`), and we can add some info there to give context.

Also a reminder to try and hard-wrap this line at 80 while you're in the file.

adellape commented on dev_guide/templates.adoc:
```diff
@@ -51,41 +51,22 @@ A template describes a set of resources intended to be used together that can be
+== Creating a Configuration From a Template
 
-There is also the ability to add labels in the template from the command line.
+You can create a configuration from a template with either the Management Console or the CLI, after first uploading a template to your OpenShift instance.
```
We should probably start off with some context on what a configuration is / its relationship with a template. Maybe at least move the later line "When OpenShift processes a template, it creates a configuration" up to here to start this para off.

For the current sentence on this line, it's maybe also more accurate to say something like: "You can create a configuration from a template using the CLI or, if templates have been uploaded to your project or global template library, using the Management Console."

bfallonf replied:
@adellape I'm not sure I understand the relevance of the first part of your comment here. The section above is the first and general context section about the relevance of templates. I've rewritten this, so it might make more sense now. Thoughts?

Second part: done.

bparees replied:
i'm not sure if it's worth getting into, but you don't have to upload the file first if you're using the CLI, you can just run:
osc process -f filename.json | osc create -f -

whereas if you want to create from an uploaded template, it's:
osc process templatename | osc create -f -

bfallonf replied:
@bparees I included the first example, but I'm not sure if I understand what you mean about the second. You say that you don't have to upload a file first, but if you want to create from an uploaded template, use that command? I think I'm missing something...

bparees replied:
Yes...  osc process can take two types of input:
1) the name of a template that has been uploaded to the openshift template library.
  - osc process templatename
2) the name of a file on the local filesystem
  - osc process -f filename.json

In the web console, you can effectively only do #1 (you have to reference a template that has been uploaded), but when using the CLI you can do #1 or #2.  

Obviously if you have a file on the filesystem you can always upload it first, and then do #1, also.

adellape commented on dev_guide/templates.adoc:
```diff
@@ -51,41 +51,22 @@ A template describes a set of resources intended to be used together that can be
-****
-`$ osc process -f _<filename>_ -l name=otherLabel`
-****
+Create a template json file, like the above example, then upload it with the the CLI using the following process:
```
s/json/JSON/

s/the the/the/

bfallonf replied:
Done

adellape commented on dev_guide/templates.adoc:
```diff
@@ -51,41 +51,22 @@ A template describes a set of resources intended to be used together that can be
-== Parameters
-The list of parameters that you can override are listed in the `*parameters*` section of the template. Alternatively, you can override them with the following command, specifying the file to be used:
+=== Uploading a Template
+Load a json template file into the OpenShift template library with the following command:
```
Suggest:

"You can upload a template to your currently active project's template library by passing link:#overview[a JSON file] to the following command:"

And maybe add an example showing that you can also specify a different project using `-n <project>`.

bfallonf replied:
@adellape 

1. Done. Although, which overview are you intending this to link to? 

2. Done

3. I've put this in as 2.4. I think that makes sense placement-wise. However, I feel like there's a step missing. Do you have to upload a new file?

bfallonf replied:
Wait, i get it. It's passing them to the overview para above. Hmm not sure about that one as it's directly above them.

adellape replied:
Just an effort to link back to an example of what the JSON could look like, for convenience. And the "Overview" happened to be where the example is. If you don't like that, you could also create your own custom anchor right above the example block, by adding something like:

```
[[sample-template-json-file]]
```

and later linking to that:

```
by passing link:#sample-template-json-file[a JSON file] to the following command:
```

adellape commented on dev_guide/templates.adoc:
```diff
@@ -94,7 +75,7 @@ Generate a configuration with the following command:
 ****
 
-Continuing with the previous example, you can override the *`ADMIN_USERNAME`* and *`MYSQL_DATABASE`* environment variables to create a JSON string to apply to the server. The JSON string represents a configuration with customized environment variables, as shown in the following example.
+You can override the *`ADMIN_USERNAME`* and *`MYSQL_DATABASE`* link:../dev_guide/templates.html#Parameters[parameters] to create a JSON string to apply to the server. The JSON string represents a configuration with customized environment variables:
```
The anchor should be lowercase `#parameters`.

adellape commented on dev_guide/templates.adoc:
```diff
@@ -51,41 +51,22 @@ A template describes a set of resources intended to be used together that can be
-
-== Generating a Configuration
+=== Generating a Configuration
 When OpenShift processes a template, it creates a configuration. You can override the parameters with the values defined in the configuration file by adding the appropriate flags.
```
If you do move this "When OpenShift processes a template, it creates a configuration" line up, as suggested above, then the 2nd sentence here could just move to after the following example, where it would align with the example about overriding parameters.

bfallonf replied:
@adellape Hmm I think I get your above comment about moving things now. That said, as the sentence is basically a repeat of the intro para, I've deleted it. As for moving the sentence about overriding parameters, I've done as you suggested and moved it down, but also rewritten the para.

adellape commented on dev_guide/templates.adoc:
```diff
@@ -105,13 +86,77 @@ $ osc process -f examples/sample-app/application-template-dockerbuild.json -v AD
+
+=== Creating from Template
+
+To create an configuration from a template using the Management Console:
```
s/an configuration/a configuration/

Maybe:

s/a template/an link:#uploading-a-template[uploaded template]/ ?

bfallonf replied:
Done

adellape commented on dev_guide/templates.adoc:
```diff
@@ -105,13 +86,77 @@ $ osc process -f examples/sample-app/application-template-dockerbuild.json -v AD
+
+To create an configuration from a template using the Management Console:
+
+1. Click on the *Create+* button:
```
Maybe mention something about first making sure you're in the right project, or choose the desired project, etc.

bfallonf replied:
Changed to "While in the desired project, click on the *Create+* button:"

adellape commented on dev_guide/templates.adoc:
```diff
@@ -105,13 +86,77 @@ $ osc process -f examples/sample-app/application-template-dockerbuild.json -v AD
+image:console_create.png["Console Create"]
+====
+
+2. Select a template from the list of templates in your project, or provided by OpenShift:
```
Maybe:

s/OpenShift/the global template library/

Side note: Ultimately we should have an `admin_guide` topic about cluster admins adding templates to the global library in the "openshift" project; then "global template library" here could link to that.

adellape commented on dev_guide/templates.adoc:
```diff
@@ -105,13 +86,77 @@ $ osc process -f examples/sample-app/application-template-dockerbuild.json -v AD
+
+The following shows the output when listing the parameters for one of the *_openshift/origin/examples_* templates.
+
+.Example Template Output
```
"Example" is extraneous here (it renders the title as "Example 4. Example Template Output"). Simply "Template Output" is not really a very helpful title, though. Maybe "Listing Available Template Parameters"... Though honestly this one could probably just lose the title altogether.

bparees replied:
this also isn't template output, it's just showing the result of running osc process with the --parameters argument which just dumps the available parameters in the template, it does not actually process the template (so it cannot be piped to osc create)

adellape commented on dev_guide/templates.adoc:
```diff
@@ -105,13 +86,77 @@ $ osc process -f examples/sample-app/application-template-dockerbuild.json -v AD
+`$ osc process --parameters -f _<filename>_`
+****
+
+The following shows the output when listing the parameters for one of the *_openshift/origin/examples_* templates.
```
`*_openshift/origin/examples_*` could be a link to the actual repo:

https://github.com/openshift/origin/tree/master/examples

bparees commented on dev_guide/templates.adoc:
```diff
@@ -105,13 +86,77 @@ $ osc process -f examples/sample-app/application-template-dockerbuild.json -v AD
-The JSON string can either be redirected to a file or applied directly by piping the process output to the `apply` command, as shown in the following example.
+The JSON string can either be redirected to a file or applied directly by piping the process output to the `create` command:
 
 .Redirecting JSON String
```
title is wrong since the body shows piping the content to osc create.

bfallonf replied:
Changed to "Applying the JSON String", but I'm not convinced that's the best wording...Hard to do it without using "The `create` command".

bparees replied:
i'd rather us avoid referring to it as "the json string" entirely if possible, that's pretty technical and generic.

"Applying the template output" or "Applying the template resources" or something.  whatever we're referring to the template as producing when processed.

bparees commented on dev_guide/templates.adoc:
```diff
@@ -105,13 +86,77 @@ $ osc process -f examples/sample-app/application-template-dockerbuild.json -v AD
+The list of parameters that you can override are listed in the `*parameters*` section of the template. Alternatively, you can override them with the following command, specifying the file to be used:
+
+****
+`$ osc process --parameters -f _<filename>_`
```
this example lists the parameters, it does not override them.  if you want to override the parameter values, you use the -v flag, as in  "osc process -v FOO=BAR,KEY=VAL -f mytemplate.json"

bparees commented on dev_guide/console.adoc:
```diff
@@ -69,34 +69,3 @@ image:console_filter_step3.png["Console Filter Step 3"]
 ====
-
-
-== Creating from Template
```
we probably still need some sort of link from the console doc to templates...  eg a high level summary of "here's all the stuff you can do in the console (link, link, link)"

bfallonf replied:
@bparees Sorry, not too sure what you mean here too. Alex mentioned above that Bilhar's suggested that we split up the console doc and put all the bits into other topics, so the Creating from a Template section will be in this section after (if) this PR is merged. 

Do you mean we should be linking to other parts of the docs to show what else can be created from templates?

bparees replied:
if we're getting rid of the console section entirely I guess that's ok (sorry I didn't catch up on all the iteration that occurred while I was out) then...but I was assuming and think it would be better if we still have a section dedicated to the console that would cover at a high level what you can do from the console.    As a user I don't want to have to read through specific sections to find out what i can do in the console.  **How** I do those things in the console belong in the specific sections.

adellape replied:
+1

bfallonf replied:
Ah ok. @bparees I see what you mean now. Including this doc as a whole might be something I can talk to @CowboysFan  about. As for now, I've added a small section to the console doc about creating from templates, linking to here. It's really a placeholder; once we settle on what else should be here it might work better as a list or table or something.

adellape commented on architecture/core_objects/overview.adoc:
```diff
@@ -5,7 +5,20 @@
+section outlines how the Kubernetes infrastructure is used in your OpenShift
+instance. The link:../architecture/core_objects/openshift-model.html[OpenShift
+Model] section outlines any OpenShift-specific information about your
+infrastructure.
```
Thanks for filling this stub out. It's a bit of a misnomer to say the `kubernetes_model` topic outlines the "Kubernetes infrastructure", when in a neighboring directory we have `infrastructure_components/kubernetes_infrastructure`. Maybe something like:

"The following topics provide information on important resources in the link:kubernetes_model.html[Kubernetes model], which is extended by the link:openshift_model[OpenShift model]."

bfallonf replied:
Changed to "The Kubernetes model topic provides information about important resources of your OpenShift instance, which is extended by the OpenShift model topic."

I think putting in "the following" might be specific to books or sections of the same topic.

adellape commented on architecture/core_objects/overview.adoc:
```diff
@@ -5,7 +5,20 @@
+A number of Kubernetes objects can be created by utilizing templates.
+
+See the link:../dev_guide/templates.html[Templates] section for more
+information.
```
You can just move this link in-line to the above sentence (also a minor rewrite suggestion/idea here):

"You can create many of these resources at once by utilizing link:../dev_guide/templates.html[templates]."

bfallonf replied:
Done

adellape commented on architecture/core_objects/overview.adoc:
```diff
@@ -5,7 +5,20 @@
+The link:../architecture/core_objects/builds.html[Builds] section outlines how
+to create a working image from a build object, and the
+link:../architecture/core_objects/routing.html[Routing] seciton outlines the
+routing process.
```
s/seciton/section/

bfallonf replied:
Oops

bfallonf commented:
@bparees @adellape I think this is it for comments. If there's anything else, please let me know. Otherwise, this might be ready to merge. Thanks all!

CowboysFan commented:
@bparees @adellape @bfallonf The Dev Guide will be task based, with those tasks categorized. If any task can be accomplished with the CLI and the console, then we will document those there. The only way I see a separate console dir/topic is if customers can accomplish every CLI task in the console as well. Otherwise, we're going to keep our docs task focused, and that will result in this topic being broken up into those relevant tasks. Hope that makes sense.

bparees commented:
@CowboysFan yeah i think so.

bparees commented on dev_guide/templates.adoc:
```diff
@@ -51,67 +56,154 @@ A template describes a set of resources intended to be used together that can be
-MYSQL_ROOT_PASSWORD   database password        expression          [a-zA-Z0-9]{8}
-MYSQL_DATABASE        database name                                root
-----
+Alternatively, you can upload a template then create from the same template by
```
that command isn't uploading the template, it's just processing it from the filesystem, so the template doesn't end up in the template library.

bfallonf replied:
Oh, now I get you. I've changed this to:

"Alternatively, you can create from a template without uploading it to the
template library by processing the template and creating from the same template
by piping both commands:"

bparees commented on dev_guide/templates.adoc:
```diff
@@ -51,67 +56,154 @@ A template describes a set of resources intended to be used together that can be
+You can override any link:../dev_guide/templates.html#parameters[parameters]
+defined in the JSON configuration file by adding the `-v` option and any desired
+parameters. For example, you can override the *`ADMIN_USERNAME`* and
+*`MYSQL_DATABASE`* parameters to create a JSON string with customized
```
i still don't like that we're calling this a "JSON string" everywhere.  the user isn't trying to create a "JSON string", they're trying to create a definition of the objects they ultimately want to create in openshift.

bfallonf replied:
Ok. I think I know what you mean. I've taken out the few references where it says JSON string, and worded around it so it sounds ok.

bfallonf commented:
Made a few more changes after @bparees 's suggestions.

adellape commented on architecture/core_objects/overview.adoc:
```diff
@@ -5,7 +5,14 @@
+routing process.
+
+You can create many of these resources by utilizing
+link:../dev_guide/templates.html[Templates].
```
Most of the links in this topic are bad.

bfallonf replied:
This took me a while, but fixed.

adellape commented on dev_guide/console.adoc:
```diff
@@ -70,33 +70,6 @@ image:console_filter_step3.png["Console Filter Step 3"]
-<4> Parameters defined by the template. You can edit values for parameters defined in the template here.
+== Creating a Configuration From a Template
+You can create many important Kubernetes and OpenShift resources from the
+Management Console by utilizing link:../dev_guide/templates.html[Templates].
```
"Templates" doesn't necessarily need to be capitalized here.

bfallonf replied:
Done

adellape commented on dev_guide/templates.adoc:
```diff
@@ -51,67 +56,153 @@ A template describes a set of resources intended to be used together that can be
 
 ****
-`$ osc process -f _<filename>_ -l name=otherLabel`
+`osc create -f _<filename>_`
```
All throughout this topic, there are many example commands missing the `$` command prompt at the beginning.

bfallonf replied:
Done

adellape commented on dev_guide/templates.adoc:
```diff
@@ -51,67 +56,153 @@ A template describes a set of resources intended to be used together that can be
-When OpenShift processes a template, it creates a configuration. You can override the parameters with the values defined in the configuration file by adding the appropriate flags.
+****
+`osc process -f _<filename.json>_ | osc create -f -`
+****
```
(Unless I missed it in the PR conversation somewhere) Is there a reason this whole part (from "Alternatively" through the syntax sidebar) is in the "Uploading a Template" section? Seems like strange placement. The second half of the next section, "Generating a Configuration", rehashes this concept and shows an actual example, so why not move this there so the syntax and example are together?

bfallonf commented:
@adellape You're right, your other comment disappeared. I placed the "Alternatively..." section to your suggested place. I believe my reasoning was mistakenly about it serving both purposes; uploading and creating. But it isn't about uploading. 

Let me know if there's anything for this PR.

adellape commented:
Thanks for the updates. Merging.
